### ISSUE TITLE
More utility functions for `Polysemy.Error`

### ISSUE THREAD

**Issue by cnr** [polysemy-research/polysemy#296]
I find myself re-defining some combinators from `Control.Exception`, most commonly:
```
try :: Member (Error e) r => Sem r a -> Sem r (Either e a)`
tryJust :: Member (Error e) r => (e -> Maybe b) -> Sem r a -> Sem r (Either b a)
handle :: Member (Error e) r => (e -> Sem r a) -> Sem r a -> Sem r a
```

are these (or any others) worth adding/PRing to polysemy?

**Comment by KingoftheHomeless** [comment]
`try`, definitely, `tryJust`, sure, why not. `handle` doesn't feel worth it. It'd be squatting a good name we could want for other purposes, and its main use case is already be covered by ``(`catch` handler) $ big chunk of code``.

Another good beginner issue: **add `try` and `tryJust` to Polysemy.Error**. Documentation for `try` can be bare-bones or non-existent. Documentation for `tryJust` should explain that the exception is rethrown if the predicate fails.

**Comment by bolt12** [comment]
I do not know if @cnr is giving this a go, but to dust off I tried to tackle this and came up with this:

```haskell
------------------------------------------------------------------------------
-- | Attempt to run a computation, returning an @'Either'@ with the
-- exception.
try :: Member (Error e) r => Sem r a -> Sem r (Either e a)
try m = catch (Right <$> m) (return . Left)
{-# INLINABLE try #-}

------------------------------------------------------------------------------
-- | Attempt to run a computation, with a provided exception function,
-- returnreturning an @'Either'@ with the exception.
tryJust :: Member (Error e) r => (e -> Maybe b) -> Sem r a -> Sem r (Either b a)
tryJust f m = catch (Right <$> m) (return . Left . fromJust . f)
{-# INLINABLE tryJust #-}

------------------------------------------------------------------------------
-- | Attempt to run a computation, with a provided exception function and
-- catch function.
catchJust :: Member (Error e) r => (e -> Maybe b) -> Sem r a -> (b -> Sem r a) -> Sem r a
catchJust ef m bf = catch m (bf . fromJust . ef)
{-# INLINABLE catchJust #-}
```

It type checks but seems to simple to be right and since I do not have much experience with complex error handling in Haskell I just want to ask if this is an appropriate solution!

**Comment by cnr** [comment]
@bolt12 you're welcome to open a PR for this

remember that `fromJust` is partial -- it doesn't handle the `Nothing` case! you'll probably want to pattern match and re-throw the error when the supplied function produces `Nothing`

**Comment by bolt12** [comment]
Awesome! Thanks for the insight! I already made the PR at #297

**Comment by KingoftheHomeless** [comment]
Addressed by #297.
